src/qobj/states.jl

```diff
@@ -118,13 +118,18 @@ Density matrix for a thermal state (generating thermal state probabilities) with
 !!! warning "Beware of type-stability!"
     If you want to keep type stability, it is recommended to use `thermal_dm(N, n, sparse=Val(sparse))` instead of `thermal_dm(N, n, sparse=sparse)`. See [this link](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-value-type) and the [related Section](@ref doc:Type-Stability) about type stability for more details.
 """
-function thermal_dm(N::Int, n::Real; sparse::Union{Bool,Val} = Val(false))
-    β = log(1.0 / n + 1.0)
-    data = [exp(-β * ComplexF64(j)) for j in 0:(N-1)]
-    if getVal(sparse)
-        return QuantumObject(spdiagm(0 => data ./ sum(data)), Operator(), N)
+function thermal_dm(N::Int, n::T; sparse::Union{Bool,Val} = Val(false)) where {T<:Real}
+    if n == 0
+        return fock_dm(N, 0; sparse)
```

Comment:
```suggestion
        return fock_dm(N, zero(T); sparse)
```

or something along this line to fix type stability for non-Float64.

Reply:
@yuyichao what kind of type stability are you referring to?

Reply:
The type stability is fixed now by not having a toplevel branch. This version of code used to have a type stability problem because `fock_dm` will (or at least this way of calling it would) only return an eltype of `ComplexF64` and that's not necessarily what the other branch does, e.g. when a Dual number of BigFloat is provided as nbar.